File: frontend/src/api/analytics.ts
import api from "./client";

export interface IPSummary {
  source_ip: string;
  geo_country: string | null;
  geo_city: string | null;
  geo_subdivision: string | null;
  whois_org: string | null;
  whois_asn: string | null;
  total_messages: number;
  report_count: number;
  failure_count: number;
}

export interface DailyVolume {
  date: string;
  total_messages: number;
  pass_count: number;
  fail_count: number;
}

export interface ClientAnalytics {
  client_slug: string;
  total_reports: number;
  total_records: number;
  total_messages: number;
  open_flags: number;
  flags_by_severity: Record<string, number>;
  top_ips: IPSummary[];
  daily_volume: DailyVolume[];
}

export interface CrossClientSummary {
  total_clients: number;
  total_reports: number;
  total_open_flags: number;
  clients: ClientAnalytics[];
}

export interface GeoDistributionEntry {
  country: string;
  messages: number;
}

export const analyticsApi = {
  client: (slug: string) =>
    api.get<ClientAnalytics>(`/clients/${slug}/analytics`).then((r) => r.data),

  crossClient: () =>
    api.get<CrossClientSummary>("/analytics").then((r) => r.data),

  geoDistribution: (slug: string, days = 30) =>
    api
      .get<GeoDistributionEntry[]>(`/clients/${slug}/analytics/geo-distribution`, { params: { days } })
      .then((r) => r.data),
};

File: core/schemas/analytics.py
from pydantic import BaseModel


class FlagSummary(BaseModel):
    flag_type: str
    severity: str
    count: int


class IPSummary(BaseModel):
    source_ip: str
    geo_country: str | None
    geo_city: str | None
    geo_subdivision: str | None
    whois_org: str | None
    whois_asn: str | None
    total_messages: int
    report_count: int
    failure_count: int


class DailyVolume(BaseModel):
    date: str   # ISO date string YYYY-MM-DD
    total_messages: int
    pass_count: int
    fail_count: int


class ClientAnalytics(BaseModel):
    client_slug: str
    total_reports: int
    total_records: int
    total_messages: int
    open_flags: int
    flags_by_severity: dict[str, int]
    top_ips: list[IPSummary]
    daily_volume: list[DailyVolume]


class CrossClientSummary(BaseModel):
    """Super-admin view across all clients."""
    total_clients: int
    total_reports: int
    total_open_flags: int
    clients: list[ClientAnalytics]